# INTRODUCTION

# BUSINESS uNDERSTANDING

# PROBLEM STATEMENT

# DATA uNDERSRTANDING
This dataset represents the median monthly home sales prices for 265 zip codes from April 1996 to April 2018 according to Zillow statistics.

A distinct zip code is represented by each row. Each record includes area details and monthly median sales prices for homes.

There are 14,723 rows and 272 variables:

* `RegionID`: Unique index, 58196 through 753844
* `RegionName`: Unique Zip Code, 1001 through 99901
* `City`: City in which the zip code is located
* `State`: State in which the zip code is located
* `Metro`: Metropolitan Area in which the zip code is located
* `CountyName`: County in which the zip code is located
* `SizeRank`: Numerical rank of size of zip code, ranked 1 through 14723
* `1996-04 through 2018-04`: refers to the median housing sales values for April 1996 through April 2018, that is 265 data points of monthly data for each zip code

# IMPORTING LIBRARIES

In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import itertools
import time
# from user_functions import *
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# READING THE DATASETS

In [23]:
zillow_df = pd.read_csv('zillow_data.csv')
zillow_df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
2         3  210400.0  212200.0  212200.0  ...   321000   320600   320200   
3         4  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400   
4         5   77300.0   77300.0   77300.0  ...   119100   119400   120000   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  
2   320400   320800   321200   321200   323000   326900   329900  
3  1291500  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

In [24]:
zillow_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14723 entries, 0 to 14722
Columns: 272 entries, RegionID to 2018-04
dtypes: float64(219), int64(49), object(4)
memory usage: 30.6+ MB


In [25]:
# Check for missing values
zillow_df.isna().sum()

RegionID         0
RegionName       0
City             0
State            0
Metro         1043
              ... 
2017-12          0
2018-01          0
2018-02          0
2018-03          0
2018-04          0
Length: 272, dtype: int64

We can see that the `Metro` column has many missing values. Further investigation may be necessary to learn to deal with this column.

# DATA CLEANING
The initial step is to investigate each column. 
1. Analyzing the `RegionID` column which is the unique indices.

In [28]:
print(f"Count of unique entries", zillow_df.RegionID.value_counts())
print(f"Number of Unique values", zillow_df.RegionID.nunique())

Count of unique entries 73724     1
70551     1
99221     1
76688     1
82829     1
         ..
71176     1
91654     1
65029     1
100380    1
98304     1
Name: RegionID, Length: 14723, dtype: int64
Number of Unique values 14723


We can see that each entry is unique values maybe describing the different zipcodes of the properties. It would thus be necessary to change this value to a string data type.

In [29]:
zillow_df.RegionID = zillow_df.RegionID.astype('string')

2. Analyzing `Region Name` column: 
Unique Zip Code, 1001 through 99901 it it would be best to cast to a string

In [30]:
zillow_df.RegionName.value_counts()

55324    1
74561    1
73538    1
31546    1
82070    1
        ..
75182    1
55343    1
1450     1
73129    1
65536    1
Name: RegionName, Length: 14723, dtype: int64

In [36]:
zillow_df.RegionName = zillow_df.RegionName.astype('string')

In [38]:
for i in range(len(zillow_df)):
    zillow_df.RegionName[i] = zillow_df.RegionName[i].rjust(5, '0')

zillow_df[['RegionName', 'State']].sort_values(by=['RegionName'])

RegionName State
5850       01001    MA
4199       01002    MA
11213      01005    MA
6850       01007    MA
14547      01008    MA
...          ...   ...
4526       99709    AK
8438       99712    AK
4106       99801    AK
8658       99835    AK
7293       99901    AK

[14723 rows x 2 columns]

We can see that the `RegionName` column has been change to the unique codes of the different regions/ ststes between 01001 to 99901

3. Analyzing `State` column

In [42]:
print(f"Count of unique entries", zillow_df.State.value_counts())
print(" ")
print(f"Number of Unique values", zillow_df.State.nunique())

Count of unique entries CA    1224
NY    1015
TX     989
PA     831
FL     785
OH     588
IL     547
NJ     502
MI     499
IN     428
NC     428
MA     417
TN     404
VA     401
MN     375
GA     345
WA     341
WI     332
MO     319
MD     317
CO     249
KS     241
AZ     230
OR     224
OK     221
SC     206
NH     199
LA     193
AL     183
IA     158
MS     153
KY     139
CT     124
UT     121
ID     110
AR     105
NV     103
ME      86
NE      83
WV      72
MT      71
HI      62
NM      60
RI      59
DE      41
ND      31
WY      31
AK      28
SD      19
DC      18
VT      16
Name: State, dtype: int64
 
Number of Unique values 51


4. Analyzing the `Metro` column which is the Metropolitan Area in which the zip code is located. 
Initial data exploration showed the presence of 1043 NaN values. It msybe be best to deal with these. Most probable appraoch would be to fill these with None meaning that there is no Metropolitan Area in the given zipcode. Perharps the presence of a Metrpolitan Area has an infleunce on the value of the 

In [45]:
# Deal with missing values
zillow_df.Metro.fillna('None', inplace=True)

In [46]:
print(f"Count of unique entries", zillow_df.Metro.value_counts())

print(f"Number of Unique values", zillow_df.Metro.nunique())


Count of unique entries None                              1043
New York                           779
Los Angeles-Long Beach-Anaheim     347
Chicago                            325
Philadelphia                       281
                                  ... 
Silver City                          1
Corinth                              1
Fredericksburg                       1
Poplar Bluff                         1
Houghton                             1
Name: Metro, Length: 702, dtype: int64
Number of Unique values 702


5. Analyzing the sales columns

In [48]:
zillow_df.isna().sum()

RegionID      0
RegionName    0
City          0
State         0
Metro         0
             ..
2017-12       0
2018-01       0
2018-02       0
2018-03       0
2018-04       0
Length: 272, dtype: int64

TIME SERIES DATA PREPROCESSINg

# EXPLORATORY DATA ANALYSIS

font = {'family' : 'normal',
        'weight' : 'bold',
        'size'   : 22}

matplotlib.rc('font', **font)

# NOTE: if you visualizations are too cluttered to read, try calling 'plt.gcf().autofmt_xdate()'!

RESHAPING TIME SERIES DATA
it would be necessary to chnge the dataframe from wide to long format

In [ ]:
def melt_data(df):
    melted = pd.melt(df, id_vars=['RegionName', 'City', 'State', 'Metro', 'CountyName'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value':'mean'})

# ARIMA MODELING

# INTERPRETING RESULTS

# CONCLUSION